src/edge_cov.rs

```diff
@@ -0,0 +1,94 @@
+use alloy_primitives::{map::DefaultHashBuilder, Address, U256};
+use core::hash::{BuildHasher, Hash, Hasher};
+use revm::{
+    interpreter::{
+        opcode::{self},
+        Interpreter,
+    },
+    Database, EvmContext, Inspector,
+};
+
+// This is the maximum number of edges that can be tracked. There is a tradeoff between performance
+// and precision (less collisions).
+const MAX_EDGE_COUNT: usize = 65536;
+
+/// An `Inspector` that tracks [edge coverage](https://clang.llvm.org/docs/SanitizerCoverage.html#edge-coverage).
+/// Covered edges will not wrap to zero e.g. a loop edge hit more than 255 will still be retained.
+// see https://github.com/AFLplusplus/AFLplusplus/blob/5777ceaf23f48ae4ceae60e4f3a79263802633c6/instrumentation/afl-llvm-pass.so.cc#L810-L829
+#[derive(Clone, Debug)]
+pub struct EdgeCovInspector {
+    /// Map of hitcounts that can be diffed against to determine if new coverage was reached.
+    hitcount: Vec<u8>,
+    hash_builder: DefaultHashBuilder,
+}
+
+impl EdgeCovInspector {
+    /// Create a new `EdgeCovInspector` with `MAX_EDGE_COUNT` size.
+    pub fn new() -> Self {
+        Self { hitcount: vec![0; MAX_EDGE_COUNT], hash_builder: DefaultHashBuilder::default() }
+    }
+
+    /// Reset the hitcount to zero.
+    pub fn reset(&mut self) {
+        self.hitcount.fill(0);
+    }
+
+    /// Get the hitcount as a byte vector.
+    pub fn get_hitcount(&self) -> &[u8] {
```

Comment:
how can this be used outside of this crate? should there be a getter that uses edge_hash ?

Reply:
The fuzzer would maintain its own history for the edges and diff against the latest one (returned by `get_hitcount`) after execution. The actually index doesn't need to be known, just whether it's an "interesting change" that should be put in the corpus. https://github.com/h0mbre/Lucid/blob/3026e7323c52b30b3cf12563954ac1eaa9c6981e/src/coverage.rs#L58-L85